File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        //vector<vector<int>>ans;
        sort(nums.begin(),nums.end());
        return ksum(0,nums.size(),nums,3,0);
        
    }
    vector<vector<int>> ksum(int index, int len,vector<int>& nums, int k,int target ){
        vector<vector<int>>ans;
        if(index >=len){
            return ans;
        }
        if(k==2){
            int i= index, j= len-1;
            while(i<j){
                if(nums[i]+nums[j] ==target){
                    vector<int>temp;
                    temp.push_back(nums[i]);
                    temp.push_back(nums[j]);
                    ans.push_back(temp);
                    while(i<len-1 && nums[i]==nums[i+1]) i++;
                    while(j>0 && nums[j]==nums[j-1]) j--;
                    i++;
                    j--;

                }else if(nums[i]+nums[j] > target){
                    j--;
                }else{
                    i++;
                }
            }
        }else{
            for(int i=index;i<len;i++){
                vector<vector<int>>cur=ksum(i+1, len,nums, k-1, target-nums[i]);
                if(!cur.empty()){
                    for(auto& res:cur){
                        res.insert(res.begin(),nums[i]);
                    }
                    for(auto v:cur){
                        ans.push_back(v);
                    }
                }
                while(i <len-1 && nums[i]==nums[i+1])i++;
            }
        }
        return ans;
    }
};

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        return kSum(nums, 0, 0, nums.length, 3);
    }
    public List<List<Integer>> kSum(int[] nums, int index, int target, int n, int k){
        List<List<Integer>> answer = new ArrayList<>();
        if(index >= n){
            return answer;
        }
        if(k==2){
            int i = index, j = n-1;
            while(i < j){
                if(nums[i]+nums[j] == target){
                    List<Integer>newList = new ArrayList<>();
                    newList.add(nums[i]);
                    newList.add(nums[j]);
                    answer.add(newList);
                    while(i < n-1 && nums[i]==nums[i+1]) i++;
                    while(j >0 && nums[j] == nums[j-1]) j--;
                    i++;
                    j--;
                }else if(nums[i]+nums[j] < target){
                    i++;
                }else{
                    j--;
                }
            }
        }else{
            for(int i= index;i<n-k+1;i++){
                List<List<Integer>>cur = kSum(nums, i+1, target-nums[i], n, k-1);
                if(cur!=null){
                    for(List<Integer>list: cur){
                        list.add(0, nums[i]);
                    }
                    answer.addAll(cur);
                }
                
                while(i < n-1 && nums[i]==nums[i+1]) i++;
            }
        }
        return answer;
    }
}